File: Ternary_Search.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    int n;
    cin >> n;
    vector<int> arr(n);
    for(int i = 0; i < n; i++) {
        cin >> arr[i];
    }
    
    int left = 0, right = n-1;
    while(right - left >= 3) {
        int m1 = left + (right - left)/3;
        int m2 = right - (right - left)/3;
        
        if(arr[m1] < arr[m2]) {
            left = m1;
        } else {
            right = m2;
        }
    }
    
    int peak = arr[left];
    for(int i = left; i <= right; i++) {
        peak = max(peak, arr[i]);
    }
    
    cout << peak << endl;
    return 0;
}

File: Ternary_Search.py
n = int(input())
arr = list(map(int, input().split()))

left = 0
right = n - 1
print(*list(range(n)), sep=" ")

while right - left >= 3:
    # Divide array into three parts
    mid1 = left + (right - left) // 3
    mid2 = right - (right - left) // 3

    # Compare values at mid points
    if arr[mid1] > arr[mid2]:
        # Peak lies in left segment
        right = mid2
    else:
        # Peak lies in right segment
        left = mid1

# Find maximum in remaining small segment
peak = max(arr[left : right + 1])
print(peak)
